### DISCUSSION TITLE
Problem with middleware and custom-server

### DISCUSSION THREAD

**Discussion by ost-ing** [vercel/next.js#33374] [Help]
Hello,

I've written some middleware using the new API to rewrite urls for fully translated routing within my app. I had it all working with a custom express server using version `12.0.5-canary.11`.

I then subsequently upgraded to `12.0.7` and now get this error:

```
Error: To use middleware you must provide a `hostname` and `port` to the Next.js Server
```

This check was introduced in [`12.0.5-canary.18` ](https://github.com/vercel/next.js/pull/31935/commits/a29a1fd624f17e92943c00b787fae2d1eaae52ce)

How can I resolve this problem in my custom server?

**Comment by iksent** [comment]
I use custom `server.js`.

Passing `hostname` and `port` params helped me:

```
...
const isDev = process.env.NODE_ENV !== 'production'

const port = Number(process.env.PORT || 3030)
const app = next({ dev: isDev, hostname: 'localhost', port })
...
```

**Reply by iksent** [reply]
But there are no information [in the docs](https://nextjs.org/docs/middleware) about this...

**Reply by fortezhuo** [reply]
@iksent you save my day.. I am struggled with this error `Error [TypeError]: Invalid URL: http://undefined:undefined/?__nextDefaultLocale=` while using custom server and middleware. And adding port and hostname solve the problems.

**Reply by pratikkapadia5** [reply]
I am also currently facing similar issue.

**Reply by hoon2hooni** [reply]
I also faced same issue but you saved me thank u ~

**Reply by robiulhr** [reply]
> I use custom `server.js`.
> 
> Passing `hostname` and `port` params helped me:
> 
> ```
> ...
> const isDev = process.env.NODE_ENV !== 'production'
> 
> const port = Number(process.env.PORT || 3030)
> const app = next({ dev: isDev, hostname: 'localhost', port })
> ...
> ```

I have implemented the server maintaining all these things and my application is working fine in the localhost but getting problems when I deploy it on the shared hosting. all the routes mentioned in the config.matcher are showing not-found page instead of its page. I have tried using the hostname for the production for example: `example.com`, `120.320.245.43` (shared IP address), but it is not working. 

for details please check this one https://github.com/vercel/next.js/discussions/68754#discussion-7032579
